persist/fs/commitlog/reader.go

```diff
@@ -56,34 +61,84 @@ type commitLogReader interface {
 	Close() error
 }
 
+type readResponse struct {
+	series     Series
+	datapoint  ts.Datapoint
+	unit       xtime.Unit
+	annotation ts.Annotation
+	resultErr  error
+}
+
+type decoderArg struct {
+	bytes checked.Bytes
+	err   error
+}
+
+type readerMetadata struct {
+	sync.RWMutex
+	decoder encoding.Decoder
+	lookup  map[uint64]Series
+	wgs     map[uint64]*sync.WaitGroup
+}
+
 type reader struct {
-	opts            Options
-	bytesPool       pool.CheckedBytesPool
-	chunkReader     *chunkReader
-	sizeBuffer      []byte
-	dataBuffer      []byte
-	logDecoder      encoding.Decoder
-	metadataDecoder encoding.Decoder
-	metadataLookup  map[uint64]Series
+	opts          Options
+	numConc       int
+	bytesPool     pool.CheckedBytesPool
+	chunkReader   *chunkReader
+	dataBuffer    []byte
+	logDecoder    encoding.Decoder
+	decoderBufs   []chan decoderArg
+	outBufs       []chan readResponse
+	cancelCtx     context.Context
+	cancelFunc    context.CancelFunc
+	shutdownChan  chan error
+	metadata      readerMetadata
+	nextIndex     int
+	hasBeenOpened bool
 }
 
 func newCommitLogReader(opts Options) commitLogReader {
 	decodingOpts := opts.FilesystemOptions().DecodingOptions()
-	return &reader{
-		opts:            opts,
-		bytesPool:       opts.BytesPool(),
-		chunkReader:     newChunkReader(opts.FlushSize()),
-		sizeBuffer:      make([]byte, binary.MaxVarintLen64),
-		logDecoder:      msgpack.NewDecoder(decodingOpts),
-		metadataDecoder: msgpack.NewDecoder(decodingOpts),
-		metadataLookup:  make(map[uint64]Series),
+	cancelCtx, cancelFunc := context.WithCancel(context.Background())
+
+	numConc := opts.ReadConcurrency()
+	decoderBufs := []chan decoderArg{}
```

Comment:
nit: you can pre-allocate here

```
decoderBufs := make([]chan decoderArg, 0, numCon)
```

Reply:
I usually try to avoid doing this when its not performance sensitive (this only happens one time in the entire bootstrapping process) because I find its a really common source of bugs, but I can change it

Reply:
Yeah not the biggest difference perf wise but might as well whilst we're here.